### PR TITLE
Add Git Clone menu option

### PR DESCRIPTION
**Pull request by jeremypw**
Partly addresses #1307 except it provides a context menu option not a keyboard shortcut.

### CODE REVIEW

File: plugins/git/plugin.vala [hunk 1]
```diff
@@ -167,6 +167,65 @@ public class Marlin.Plugins.Git : Marlin.Plugins.Base {
             }
         }
     }
+
+    public override void context_menu (Gtk.Widget? widget, List<GOF.File> files) {
+        var menu = (Gtk.Menu)widget;
```
Comment (tintou) [inline plugins/git/plugin.vala:172]:
```suggestion
        unowned Gtk.Menu? menu = (Gtk.Menu)widget;
```

**Comment by jeremypw** [comment]
@tintou Do you know how to have a spinner and cancel button work when Ggit.Repository.clone () is synchronous and does not take a cancellable as a parameter?  Is it even possible?

**Comment by marbetschar** [comment]
Any chance to get this merged? I'd LOVE to see it in Files 😍️

**Comment by jeremypw** [comment]
Looking into using GLib.Task to make the cloning cancellable.

**Comment by jeremypw** [comment]
Looks like you cannot actually cancel the cloning operation using GTask, as far as I can see.  You can "cancel" the task but the cloning continues to run in its own thread so the files are still downloaded.  But at least the Files UI is not blocked.

**Comment by jeremypw** [comment]
I think there is a way to effectively cancel the cloning but it is a bit of a hack.  If you trash the ".git" folder then cloning ends with an error (!).

**Comment by jeremypw** [comment]
One question to be decided is whether, when cloning into an empty folder from the background menu, a new root folder should be created with named from the git repo.  At the moment cloning into a location creates the .git in that location, i.e. that folder becomes the root folder.  It is therefore left to the user to create a suitable empty root folder first.  It might be convenient for Files to do that automatically in this circumstance (and not when cloning from the selection context menu).

Probably should consider having a similar feature in Code.

**Comment by marbetschar** [comment]
I guess it would be best to do the same as git clone does when executed as terminal command: create an empty subfolder using the xyz filename from the url: GitHub.com/group/xzy.git

**Comment by jeremypw** [comment]
> I guess it would be best to do the same as git clone does when executed as terminal command: create an empty subfolder using the xyz filename from the url: GitHub.com/group/xzy.git

Yes, it is a bit annoying that `GGit.Repository.clone (origin, target, options)` does not already do this.

**Comment by jeremypw** [comment]
I don't think it is possible to implement git cloning using `<Ctrl>V` of a git url without moving a lot of git stuff out of the plugin into libcore.  Do we really want to add a dependence on libgit2-glib to the core?

**Comment by marbetschar** [comment]
If the context menu action does the job but without keyboard shortcut, I think we are already 90% there - and the shortcut isn’t too important at all in my opinion.

So if it works as context action and without forcing further dependencies to core it seems good enough to me.

What do you think?

**Comment by jeremypw** [comment]
I agree.  I suspect that it would be much more difficult to get it merged if it involved substantial changes to the core.

**Comment by marbetschar** [comment]
@jeremypw any chance you have some time to resolve the conflicts, so this PR can be re-reviewed? 🤗️

**Comment by jeremypw** [comment]
@marbetschar I have resolved the conflicts and also allowed the plugin to clone local repositories as well.

**Comment by marbetschar** [comment]
@jeremypw just installed this on Odin and it seems to have some problems with detecting Git repositories for SSH urls (just put the URLs into clipboard using the copy button on GitHub while in Firefox):

```
** (io.elementary.files:35374): WARNING **: 14:08:31.433: plugin.vala:291: Not a git repository - git@github.com:marbetschar/time-limit.git

** (io.elementary.files:35374): WARNING **: 14:08:44.063: plugin.vala:291: Not a git repository - git@github.com:elementary/files.git
```

FWIW: The HTTP URLs do work without issues. Although the text (or icon) alignment for the progress indication seems to be off vertically:

Screenshot from 2021-02-08 14-12-00

**Comment by jeremypw** [comment]
@marbetschar Thanks for the further testing - I must admit I had not considered SSH urls till now.  I'll see if I can implement them.

**Comment by jeremypw** [comment]
Looks like GitHub omits the `ssh://` protocol - which explains why it does not work.  I guess I could have the cloner assume the protocol is SSH if the url begins in the form `[user@]host.xz[:port]` (?)

**Comment by jeremypw** [comment]
Another thing that this PR does not address if where the `git clone` command requires user feedback.

**Comment by jeremypw** [comment]
@marbetschar I have now set up so that I can clone GitHub repos using SSH in the terminal successfully.  However, this PR, using the libgit2 package,  gives an error "Authentication required but no callback provided".  There does not seem to be a simple solution to this (may be a bug in the package) as there is no parameter for a callback in the `Ggit.Repository.clone ()` command.

Would you consider it worthwhile merging a functionality limited cloning local and https urls for now and addressing ssh later?

**Comment by jeremypw** [comment]
I have removed the code trying to vaildate the URL.  It is now just passed to the `Ggit.Repository.clone ()` and an error shown if anything fails.

**Comment by jeremypw** [comment]
The alignment problem seems to be a style sheet issue - I cannot find anything in the code that would be causing this.

**Comment by marbetschar** [comment]
@jeremypw

> Looks like GitHub omits the ssh:// protocol - which explains why it does not work. I guess I could have the cloner assume the protocol is SSH if the url begins in the form [user@]host.xz[:port] (?)

Yes, from what I can tell this should be Ok. Maybe its fine to just check, if the URL ends with `.git` - and then, if no specific protocol is given, use SSH. Protocol wise, thats how I'd expect `git clone` works internally.

> Another thing that this PR does not address if where the git clone command requires user feedback.

I understand. Usually that's just confirming the servers' fingerprint on initial connection. Of course this would be nice to do so in the GUI with a yes/no dialog - but throwing an appropriate message is probably fine for now.

> However, this PR, using the libgit2 package, gives an error "Authentication required but no callback provided". There does not seem to be a simple solution to this (may be a bug in the package) as there is no parameter for a callback in the `Ggit.Repository.clone ()` command. Would you consider it worthwhile merging a functionality limited cloning local and https urls for now and addressing ssh later?

Personally, for 9 out of 10 repositories I use the SSH protocol to clone. Otherwise I'm not able to make any modifications, because HTTPS is anonymous therefore I'm not authenticated - which implies I'm not authorized to make any changes.

Just a wild guess: Is it possible to pass the path to the default public key `~/.ssh/id_*.pub` as option? Maybe thats the missing piece? (usually `id_rsa.pub`, but [GitHub changed this to `id_ed25519.pub`](https://docs.github.com/en/github/authenticating-to-github/adding-a-new-ssh-key-to-your-github-account))

If this does not work, my next best guess is the VAPI might be missing something. Unfortunately valadoc.org seems to be down atm, but maybe @tintou can help out regarding the missing callback argument in `Ggit.Repository.clone ()` of `libgit2`?

**Comment by jeremypw** [comment]
Looks like the the vapi does have ssh credential related stuff - just need to work out how to use it ...

**Comment by jeremypw** [comment]
Looks like the required callback is passed in the FetchOptions of the CloneOptions parameter of the clone functions.  There are some clues here https://gist.github.com/zchee/444c8c20aa7756468d8e.

**Comment by marbetschar** [comment]
@jeremypw thanks for the hint! Just made some small adjustments - now cloning using SSH works as well as cloning with HTTP.

However, the context menu option now seems always to be shown regardless of whats exactly is in the clipboard. No clue why ;)

**Comment by jeremypw** [comment]
@marbetschar  Thanks!  I tried much the same thing but for some reason it did not work for me.  I'll try your version.

**Comment by jeremypw** [comment]
I can confirm that I can now clone repositories from GtiHub with Files using SSH :tada:  

I'll look at the context menu issue now.

**Comment by jeremypw** [comment]
Very basic validation added for clipboard contents.  Providing a comprehensive Regex for valid urls of git repositories is non-trivial. Leave for another PR?

**Comment by marbetschar** [comment]
@jeremypw works like a charm! IMHO your git repository url verification is perfectly fine (at least thats exactly how I would do it too).

IMHO this RP is ready to be merged. Maybe someone from @elementary/ux can do a review?

**Comment by danrabbit** [comment]
I'm kind of hesitant about this because I feel like we might be getting into kind of a scope creep/feature overlap situation with Code. When we're working with git repositories, it seems to me like the expected workflow would be to open Code to create a new project or clone an existing project

**Comment by jeremypw** [comment]
Yes, I think you are right that this should be implemented in Code first (although doing it with Files makes it easy to specify where to create the clone.  I'll close this for now and look at doing it in Code.

**Comment by marbetschar** [comment]
@danrabbit / @cassidyjames I'm aware of the feature overlap situation with Code and I do understand the concerns of feature boilerplate creeping into Files: Basically everything we do on a Computer is somehow Files related - so a line needs to be drawn somewhere. Just wanted to check back if the decision to explicitly _not_ include Git repo cloning is final?

My personal use case: I don't use Code for all projects. Depending on the project there are better suited editors/IDE's. But all of these projects are stored in a Git repo somewhere - and all of them need to be cloned somewhen as first step. So that's the least common divisor. From there on, its very project specific what comes next.

I can use the Terminal for all of these stuff for sure - so its not a deal breaker. It would just have been super convenient to quickly clone from Files.

**Pull request closed without merging**